aten/src/ATen/TensorImpl.cpp

```diff
@@ -59,58 +59,76 @@ void Tensor::backward(
   pImpl->backward(std::move(gradient), keep_graph, create_graph);
 }
 
-TensorImpl::TensorImpl(TensorTypeId type_id, ScalarType scalar_type)
-    : type_id_(type_id), scalar_type_(scalar_type) {
-  auto type = &globalContext().getType(tensorTypeIdToBackend(type_id), scalar_type);
-  Storage* storage = type->storage(true).release();
-  StorageImpl* storage_impl = storage->pImpl();
-  tensor = new THTensor(storage_impl);
+TensorImpl::TensorImpl(TensorTypeId type_id, ScalarType scalar_type, bool is_variable)
+    : TensorImpl(nullptr, type_id, scalar_type, is_variable) {
+  // UndefinedTensors and SparseTensors don't have storages.
+  if (type_id != UndefinedTensorId() && scalar_type != ScalarType::Undefined
+      && type_id != SparseCPUTensorId() && type_id != SparseCUDATensorId()) {
+    auto type = &globalContext().getType(tensorTypeIdToBackend(type_id), scalar_type);
+    auto storage = type->storage(true);
+    storage_ = storage->pImpl();
+    storage_->retain();
+  }
 }
 
+TensorImpl::TensorImpl(StorageImpl* storage, TensorTypeId type_id, bool is_variable)
+    : TensorImpl(storage, type_id, storage->scalar_type(), is_variable) {}
+
+TensorImpl::TensorImpl(StorageImpl* storage, TensorTypeId type_id, ScalarType scalar_type, bool is_variable)
+    : storage_(storage),
+      storage_offset_(0),
+      sizes_{0},
+      strides_{1},
+      type_id_(type_id),
+      scalar_type_(scalar_type),
+      is_variable_(is_variable) {}
+
 TensorImpl::~TensorImpl() {
-  if (tensor) tensor->release();
+  if (storage_) {
+    storage_->release();
+    storage_ = nullptr;
+  }
 }
 
 IntList TensorImpl::sizes() const {
-  // NB: dim in tensor is not synchronized with THTensor, so it's
-  // important to apply dim here
-  return IntList(THTensor_getSizePtr(tensor), dim());
+  return sizes_;
 }
 
 IntList TensorImpl::strides() const {
-  // NB: dim in tensor is not synchronized with THTensor, so it's
-  // important to apply dim here
-  return IntList(THTensor_getStridePtr(tensor), dim());
+  return strides_;
 }
 
 void TensorImpl::release_resources() {
-  if (tensor) {
-      tensor->release();
-      tensor = nullptr;
+  if (storage_) {
+    storage_->release();
```

Comment:
Don't you need to `storage_ = nullptr` here? Otherwise you'll free it again in the destructor.

Reply:
Yes, good catch.  I only fixed one place from the original PR :).